### PR TITLE
Playbook to forward logs to external splunk instance

### PR DESCRIPTION
**Pull request by sabre1041**
@redhat-cop/cant-contain-this

### CODE REVIEW

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+* Source name override: OCP
+* Source Type: _json
+
+NOTE: It is beyond the scope of this document for the creation of custom link:https://docs.splunk.com/Splexicon:Sourcetype[sourcetype's] to perform advanced parsing of the data received.
+
+Complete the necessary fields to finalize the addition of the new TCP input
+
+NOTE: Open applicable firewall ports on the Splunk machine to allow for data to be received 
+
+Since the fluent-plugin-splunk-ex plugin sends data to Splunk in batches, Splunk must be configured separate messages based on the presence of a new line character.
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: add the word 'for'  after configured in this sentence.  example:

..., Splunk must be configured **for** separate messages based on...

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+Configure the _/opt/splunk/apps/search/local/props.conf_ file with the following content which will configure messages received from the _OCP_ source to split messages correctly:
+
+[source]
+----
+[source::OCP]
+SHOULD_LINEMERGE = false
+----
+
+Restart Splunk to apply the changes
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add colon after changes

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+==== Install Fluentd
+
+Fluentd can be installed on a RHEL machine by running the following command:
+
+[source]
+----
+curl -L https://toolbelt.treasuredata.com/sh/install-redhat-td-agent2.sh | sh
+----
+
+With the installation successfully completed, enable the *td-agent* service to be started at boot and start the service
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - colon after service

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+----
+
+==== Install Fluentd Plugins
+
+Two Fluentd plugins must be installed to receive and forward messages from OpenShift's aggregated logging framework to splunk:
+
+* Secure forward plugin
+* Fluentd extended Splunk plugin
+
+Execute the following commands to install the two plugins
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add colon after plugins

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+Since the secure forward plugin makes use of SSL as its primary transport mechanism, certificates must be configured to secure the communication channel between OpenShift and the standalone Fluentd instance. There are multiple ways in which the certificates can be provided. The most common methods include using a signed certificate from a trusted public certificate authority (CA) or to generate certificates using a private CA.
+
+For demonstration purposes, a private CA will be utilized. The plugin includes a tool called _secure-forward-ca-generate_ to generate the certificate and private key. Execute the following command to generate new certificates in a folder called _/etc/td-agent/certs/_ using a password protected private key with the value _ocpsecureforward_. 
+
+[source]
+----
+/opt/td-agent/embedded/bin/secure-forward-ca-generate /etc/td-agent/certs/ ocpsecureforward
+----
+
+CAUTION: Be sure to specify a unique private key password for your own environment
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add period to end of sentence

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+<2> A shared value between the sender and the receiver
+<3> Location of the previously generated certificate
+<4> Location of the previously generated private key
+<5> Private key passphrase
+<6> Hostname or IP of the Splunk instance
+<7> Port number of the Splunk input configured to accept messages
+<8> Format in which messages are sent to Splunk
+
+
+NOTE: To allow messages to be received on port 24284 for the secure forward plugin, ensure the proper firewall configurations are in place
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add period at end of sentence

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+Restart the _td-agent_ service to apply the changes
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add colon after changes

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+[source]
+----
+oc patch secrets/logging-fluentd --type=json --patch "[{'op':'add','path':'/data/external_ca_cert.pem','value':'$(base64 /path/to/your_ca_cert.pem)'}]"
+oc patch secrets/logging-fluentd --type=json --patch "[{'op':'add','path':'/data/external_ca_key.pem','value':'$(base64 /path/to/your_private_key.pem)'}]"
+----
+
+[NOTE]
+====
+The names external_ca_cert.pem and external_ca_key.pem provided in the code sample are the key names configured within the patched secret
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add period to end of sentence

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+The base64 utility must be present on the machine executing the commands
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add period to end of sentence

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+====
+
+
+==== Configuring Fluentd
+
+A link:https://docs.openshift.com/container-platform/latest/dev_guide/configmaps.html[ConfigMap] called _logging-fluentd_ is configured within the aggregated logging framework to specify the values to containing the configurations for Fluentd. Multiple files are containing within the ConfigMap, including a file called _secure-forward.conf_. The contents of this file are commented out as secure forward plugin is not enabled by default.
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: in first sentence "...specify the values to containing the configurations for Fluentd".  I think removing the work 'to' makes the sentence read better.

And in the second sentence, update 'containing' to 'contained':  "Multiple files are **contained** within..."

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+    </server>
+----
+<1> Value shared between both ends of the secure forward plugin
+<2> Location of the certificate used by the secure forward plugin
+<3> Location of the private key used by the secure forward plugin
+<4> Privae key passphrase
+<5> Hostname or IP address of the standalone Fluentd instance
+<6> Port number of the standalone Fluentd instance
+
+Note: It is recommended the enable_strict_verification option be uncommented to increase security between each endpoint of the secure forward plugin. This FQDN of the target instance to match the value configured in the certificates used to secure the communication channel.
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - make "Note" bold

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+
+==== Applying the changes
+
+Finally, since a portion of the Fluentd configuration involved the modification of secrets, the existing Fluentd pods will need to be deleted in order for them to make use of the updated values.
+
+Execute the following command to remove the existing Fluentd pods
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add colon after pods

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+[source]
+----
+oc delete pod -l component=fluentd
+----
+
+The DaemonSet will automatically start pods on the nodes in which they were previously deleted. 
+
+== Verification
+
+At this point, messages captured by OpenShift integrated logging solution should now be sent to Splunk and available within the Splunk user interface.
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add 'the' before or an apostrophe after the word OpenShift

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+
+The following steps can be used to verify the integration between OpenShift and Splunk using the secure forward plugin
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: Change the 'steps' to 'sections' and add a period to end of sentence.

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+
+=== OpenShift Fluentd
+
+The communication between the Fluentd pods running within OpenShift and the standalone Fluentd instance can be validated by viewing the logs in any one of the running pods. 
+
+Locate a running Fluentd pod within the project containing the logging infrastructure
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add colon to end of sentence

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+----
+
+View the logs of one of the running containers:
+
+[source]
+----
+oc logs logging-fluentd-9z0ye
+----
+
+A result similar to the following indicates there are no communication issues between OpenShift and the standalone instance of Fluentd
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add colon to end of sentence

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+
+[source]
+----
+2017-02-05 08:48:38 -0500 [info]: reading config file path="/etc/fluent/fluent.conf"
+----
+
+=== Standalone Fluentd
+
+The standalone instance of Fluentd can be validated by viewing the systemd journal for the _td-agent_ service. The following indicates no issues can be seen within Fluentd
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: nit - add colon to end of sentence

File: playbooks/operationalizing/secure-forward-splunk.adoc [hunk 18]
```diff
@@ -0,0 +1,318 @@
+---
+---
+= Forwarding Logs to Splunk Using Fluentd Secure Forward
+Andrew Block <ablock@redhat.com>
+v1.0, 2017-02-05
+:scripts_repo: https://github.com/rhtconsulting/rhc-ose
```
Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: is this the correct scripts_repo?

Comment (sabre1041) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: We should determine `scripts_repo` across all playbooks in general and update accordingly

Comment (JaredBurck) [inline playbooks/operationalizing/secure-forward-splunk.adoc]: Agreed. Will create a new issue for this.

**Comment by sabre1041** [comment]
@JaredBurck addressed issues. Please review

**Comment by JaredBurck** [comment]
@sabre1041 LGTM

**Pull request merged**